moveit_ros/perception/handeye_calibration_solver/src/handeye_solver_default.cpp

```diff
@@ -0,0 +1,373 @@
+ *  Copyright (c) 2019,  Intel Corporation.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of Willow Garage nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Yu Yan */
+
+#include <moveit/handeye_calibration_solver/handeye_solver_default.h>
+
+namespace moveit_handeye_calibration
+{
+const std::string LOGNAME = "handeye_solver_default";
+
+void HandEyeSolverDefault::initialize()
+{
+  solver_names_ = { "Daniilidis1999", "ParkBryan1994", "TsaiLenz1989" };
+  camera_robot_pose_ = Eigen::Isometry3d::Identity();
+}
+
+const std::vector<std::string>& HandEyeSolverDefault::getSolverNames()
+{
+  return solver_names_;
+}
+
+const Eigen::Isometry3d& HandEyeSolverDefault::getCameraRobotPose()
+{
+  return camera_robot_pose_;
+}
+
+bool HandEyeSolverDefault::solve(const std::vector<Eigen::Isometry3d>& effector_wrt_world,
+                                 const std::vector<Eigen::Isometry3d>& object_wrt_sensor, SENSOR_MOUNT_TYPE setup,
+                                 const std::string& solver_name)
+{
+  // Check the size of the two sets of pose sample equal
+  if (effector_wrt_world.size() != object_wrt_sensor.size())
+  {
+    ROS_ERROR_STREAM_NAMED(LOGNAME, "The sizes of two input pose samples must be equal");
+    return false;
+  }
+
+  char program_name[7] = "python";
+  Py_SetProgramName(program_name);
+  static int run_times{ 0 };
+  if (run_times == 0)  // Numpy can only be loaded once, otherwise will segfault
```

Comment:
Changed this variable to Boolean `numpy_loaded`. Yes, seems `Py_Initialize()` and `Py_Finalize()` can only be called once in any code with numpy extension. Some discussions about this can be found in this [link](https://stackoverflow.com/questions/7676314/py-initialize-py-finalize-not-working-twice-with-numpy).